consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftController.java

```diff
@@ -0,0 +1,186 @@
+    this.ibftFinalState = ibftFinalState;
+    this.ibftBlockHeightManagerFactory = ibftBlockHeightManagerFactory;
+    this.futureMessages = futureMessages;
+  }
+
+  public void start() {
+    startNewHeightManager(blockchain.getChainHeadHeader());
+  }
+
+  public void handleMessageEvent(final IbftReceivedMessageEvent msg) {
+    handleMessage(msg.getMessageData());
+  }
+
+  private void handleMessage(final MessageData messageData) {
+    switch (messageData.getCode()) {
+      case IbftV2.PROPOSAL:
+        final SignedData<ProposalPayload> proposalMsg =
+            ProposalMessage.fromMessage(messageData).decode();
+        if (processMessage(proposalMsg, messageData)) {
+          currentHeightManager.handleProposalMessage(proposalMsg);
+        }
+        break;
+
+      case IbftV2.PREPARE:
+        final SignedData<PreparePayload> prepareMsg =
+            PrepareMessage.fromMessage(messageData).decode();
+        if (processMessage(prepareMsg, messageData)) {
+          currentHeightManager.handlePrepareMessage(prepareMsg);
+        }
+        break;
+
+      case IbftV2.COMMIT:
+        final SignedData<CommitPayload> commitMsg = CommitMessage.fromMessage(messageData).decode();
+        if (processMessage(commitMsg, messageData)) {
+          currentHeightManager.handleCommitMessage(commitMsg);
+        }
+        break;
+
+      case IbftV2.ROUND_CHANGE:
+        final SignedData<RoundChangePayload> roundChangeMsg =
+            RoundChangeMessage.fromMessage(messageData).decode();
+        if (processRoundMessage(roundChangeMsg)) {
+          currentHeightManager.handleRoundChangeMessage(roundChangeMsg);
+        }
+        break;
+
+      case IbftV2.NEW_ROUND:
+        final SignedData<NewRoundPayload> newRoundMsg =
+            NewRoundMessage.fromMessage(messageData).decode();
+        if (processRoundMessage(newRoundMsg)) {
+          currentHeightManager.handleNewRoundMessage(newRoundMsg);
+        }
+        break;
+
+      default:
+        throw new IllegalArgumentException(
+            "Received message does not conform to any recognised IBFT message structure.");
+    }
+  }
+
+  public void handleNewBlockEvent(final NewChainHead newChainHead) {
+    startNewHeightManager(newChainHead.getNewChainHeadHeader());
+  }
+
+  public void handleBlockTimerExpiry(final BlockTimerExpiry blockTimerExpiry) {
+    currentHeightManager.handleBlockTimerExpiry(blockTimerExpiry.getRoundIndentifier());
+  }
+
+  public void startNewHeightManager(final BlockHeader parentHeader) {
+    currentHeightManager = ibftBlockHeightManagerFactory.create(parentHeader);
+    currentHeightManager.start();
+    final long newChainHeight = currentHeightManager.getChainHeight();
+    futureMessages.getOrDefault(newChainHeight, emptyList()).forEach(this::handleMessage);
+    futureMessages.remove(newChainHeight);
+  }
+
+  public void handleRoundExpiry(final RoundExpiry expired) {
+    currentHeightManager.roundExpired(expired);
```

Comment:
ChainHeight Check, discard if old?
maybe with logs if it's not for the current height?

Reply:
I think logging is good idea. I'll add some